### PR TITLE
Bump version to 1.47 

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
This also bumps to a more recent rustfmt version, just to keep us relatively up to date (though almost nothing has changed in rustfmt we use beyond bumps to the parser infra). No formatting changes as a result of this.

r? @pietroalbini

### CODE REVIEW

**Comment by Manishearth** [comment]
The build isn't able to find libz, could it be this PR? https://github.com/rust-lang/rust/pull/74407#issuecomment-659585142

**Comment by Mark-Simulacrum** [comment]
```
/home/mark/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustc:
        linux-vdso.so.1 (0x00007ffeec130000)
        librustc_driver-a07dcbb4ed0bdde8.so => /home/mark/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-a07dcbb4ed0bdde8.so (0x00007f858d5f7000)
        libstd-c147cd9c030850ef.so => /home/mark/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/../lib/libstd-c147cd9c030850ef.so (0x00007f858d2f6000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f858d2ca000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f858d2c4000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f858d2b9000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f858d0c7000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f8591b6a000)
        libLLVM-10-rust-1.45.0-stable.so => /home/mark/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-10-rust-1.45.0-stable.so (0x00007f8588aca000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f8588aaf000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f8588960000)
/home/mark/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/bin/rustc:
        linux-vdso.so.1 (0x00007ffd5ddb6000)
        librustc_driver-e231ac240cfaf689.so => /home/mark/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/bin/../lib/librustc_driver-e231ac240cfaf689.so (0x00007f4a387a4000)
        libstd-d66e0ceb8eaec9b8.so => /home/mark/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/bin/../lib/libstd-d66e0ceb8eaec9b8.so (0x00007f4a384a1000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f4a38475000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f4a3846f000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f4a38464000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f4a38272000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f4a3cdd2000)
        libLLVM-10-rust-1.46.0-beta.so => /home/mark/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/bin/../lib/../lib/libLLVM-10-rust-1.46.0-beta.so (0x00007f4a33c07000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f4a33bec000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f4a33a9d000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f4a33a81000)
```

Indeed it looks like we picked up a libz dependency somewhere along the line since last beta. Going to try and bisect it

**Comment by Mark-Simulacrum** [comment]
searched nightlies: from nightly-2020-06-19 to nightly-2020-07-11
regressed nightly: nightly-2020-06-22
searched commits: from https://github.com/rust-lang/rust/commit/f455e46eae1a227d735091091144601b467e1565 to https://github.com/rust-lang/rust/commit/a8cf3991177f30694200002cd9479ffbbe6d9a1a
regressed commit: https://github.com/rust-lang/rust/commit/349f6bfb11d73ebb6a272f9a3d00883484f8218c - https://github.com/rust-lang/rust/pull/72696

That is an unexpected consequence of that PR, I would've expected zlib to be statically linked to LLVM. cc @jethrogb

**Comment by Mark-Simulacrum** [comment]
cc @cuviper @rust-lang/release -- do we know if it's a problem if rustc starts requiring zlib as a dynamically linked dependency? It looks like that's what is in current nightly/beta. If it's not a problem then we can likely just add it to our i686-gnu builders (and whatever others need it)

**Comment by jonas-schievink** [comment]
The new zlib dependency also broke Rust on NixOS, so it looks like it had bigger fallout than anticipated. Might have been a good idea to put it in the release notes.

**Comment by Mark-Simulacrum** [comment]
> The new zlib dependency also broke Rust on NixOS, so it looks like it had bigger fallout than anticipated. Might have been a good idea to put it in the release notes.

To be clear we still can, this is about 1.46 beta -- 1.45 stable released today does not have this problem.

It's also not the case that we knew that we were adding a new dependency (as mentioned, I expected it to be statically linked or so). I guess we should maybe add some rudimentary checking in CI for that.

**Comment by cuviper** [comment]
> do we know if it's a problem if rustc starts requiring zlib as a dynamically linked dependency?

I thought this was already a known consequence of #72696. I expect this library is common enough to be usually assumed present, but 32-bit on 64-bit hosts might not have it. Nix is kind of special too, but they've already worked around that.

> we can likely just add it to our i686-gnu builders

The centos:5 image adds `zlib-devel`, which I think already pulls in both 32-bit and 64-bit libs, but `zlib-static` would be separate. I did add the debian equivalent manually in #74163.

**Comment by jonas-schievink** [comment]
> To be clear we still can, this is about 1.46 beta -- 1.45 stable released today does not have this problem.

Ah, right!

**Comment by mati865** [comment]
> The centos:5 image adds zlib-devel, which I think already pulls in both 32-bit and 64-bit libs

@cuviper on Fedora that will pull only x86_64 zlib, is CentOS any different here? I think zlib is built in the docker image via scripts on Rust's CI.

**Comment by cuviper** [comment]
> on Fedora that will pull only x86_64 zlib, is CentOS any different here?

Back in RHEL/CentOS 5, yum would install both by default, but 6 and later does not. I'm not sure exactly _how_ that changed, but you can explicitly explicity request `zlib-devel.i686` if needed.

However, it doesn't look like `zlib-static` is available as multilib, and I'm not sure how to trick LLVM into using it anyway. We might have to drop that devel package and manually build our own static-only library.

**Comment by Manishearth** [comment]
@bors rollup=iffy

CI changes

File: src/ci/docker/host-x86_64/i686-gnu-nopt/Dockerfile [hunk 1]
```diff
@@ -11,6 +11,7 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
   cmake \
   sudo \
   gdb \
+  zlib1g-dev \
```
Comment (cuviper) [inline src/ci/docker/host-x86_64/i686-gnu-nopt/Dockerfile:14]: I think you need `lib32z1-dev` for the i686 bits.

**Comment by Mark-Simulacrum** [comment]
I'd personally be down for reverting the PR adding zlib and pushing that into beta as well, though I imagine that's a problem for anyone trying to use lld as a linker with C compilers (which apparently need zlib?).

@bors try

**Comment by Mark-Simulacrum** [comment]
I would like to get this landed sooner rather than later, so I've filed https://github.com/rust-lang/rust/issues/74420 to discuss the zlib change further. @bors r=pietroalbini

**Comment by Manishearth** [comment]
@bors p=1

are you okay with rollup=iffying this? That might be better

**Comment by Mark-Simulacrum** [comment]
It would work, but I'm not confident I caught all of the builders that need the CI changes here. Up to you.

**Comment by Manishearth** [comment]
@bors rollup=iffy

I'm probably just going to let this go directly through, but we'll see.

**Comment by eddyb** [comment]
I wonder if we should have a NixOS builder or something. This PR is missing a change to the Python side of `src/bootstrap` for the zlib dependency, I'll try to maybe get in a fix for it later today. cc @nagisa @davidtwco (not sure who else will hit this)

**Pull request merged**